gui: Add copy to clipboard, share by email, and share by SMS buttons to device IDs (fixes #2771, ref #3868)

tomasz1986 opened the pull request:
Add buttons to allow for simpler sharing device IDs with others. The
first one copies the ID to clipboard (trying to use three different
methods, depending on the browser). The second one triggers a mailto
link with prefilled subject and body. The third one triggers an sms link
with prefilled body. The short description of Syncthing included in the
official website https://syncthing.net.

Issue #3868 is referred here, because the copy to clipboard button
offers an alternative method for IE11 users to actually be able to copy
device IDs without having to select it manually (which doesn't work).

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### Screenshots

![image](https://user-images.githubusercontent.com/5626656/190706169-88096abe-5ed3-469d-91a5-5066077f7d3a.png)

![image](https://user-images.githubusercontent.com/5626656/190706282-4ff24781-e5e9-4e89-aa8e-c67495821a1b.png)

![image](https://user-images.githubusercontent.com/5626656/190706296-5dc842fa-c32e-4e23-ab6e-98c26c18f6dc.png)

![image](https://user-images.githubusercontent.com/5626656/190706309-aa1550df-5dc3-4d10-a41d-93eb887d4584.png)

tomasz1986 commented:
This is a draft, because I still need to figure out some CSS regarding small resolutions. Right now the buttons don't wrap and overflow the screen on mobile. I'm going to add screenshots and other details once I've fixed this problem.

As far as the email text goes, I've got some inspiration from the linked issue, and I've also checked what Dropbox says in their invitation mails (which are actually quite short, probably assuming that people already know what Dropbox is). If you've got any ideas how the text could be improved, please let me know in advance. I've deliberately decided not to include any direct links straight to the Docs and such, because the receiver may be using the Android app or another wrapper that has its own UI, so I've decided to link to the official site only (which has all the required information anyway).

acolomb reviewed (changes requested):
Some GUI styling issues can probably be avoided entirely.  If you describe what different behavior you'd like to achieve with the CSS, maybe I can help find a solution.

tomasz1986 commented on gui/default/index.html:
```diff
@@ -958,7 +958,6 @@ <h4 class="panel-title">
   <script type="text/javascript" src="syncthing/core/notificationDirective.js"></script>
   <script type="text/javascript" src="syncthing/core/pathIsSubDirDirective.js"></script>
   <script type="text/javascript" src="syncthing/core/popoverDirective.js"></script>
-  <script type="text/javascript" src="syncthing/core/selectOnClickDirective.js"></script>
```
I've actually found it very annoying, e.g. when I was reporting something on the forum and wanted to copy only a part of the ID, having to fight with the automatic selection, eventually ending up temporarily disabling JS or using developer tools just to be able to copy the text.

This script isn't used in any other places, is it? With a copy to clipboard button present, is there any other real use case for it?

acolomb replied:
I'm working on a Linux desktop OS.  There is not only one clipboard, but whatever text is selected gets copied automatically to the so-called "Primary Selection" buffer.  I can instantly paste it in any other text entry field by a single middle-button click.  Two clicks overall, very efficient, no searching for a paste button or switching to the keyboard.  Same goes for the drag-and-drop usage pattern.

Of course the same can be achieved by a left button triple-click usually to select the whole paragraph / text span.  I don't think the current behavior is a must-have, just that there is no real reason to *not* keep it.



> wanted to copy only a part of the ID, having to fight with the automatic selection, eventually ending up temporarily disabling JS or using developer tools just to be able to copy the text.

That sounds awkward indeed.  Why not cut it down to what you need after pasting?

tomasz1986 replied:
> That sounds awkward indeed. Why not cut it down to what you need after pasting?

Yeah, it'd probably be easier that way. I guess it's just the fact of being prevented from selecting parts of the text that makes it feel wrong. I'd say it has a similar feeling to those sites that use JS to block any text selection. I even have a browser extension that pauses JS on one click specifically to override such behaviour.

Another reason against it would be that the current script isn't compatible with IE11, preventing any text selection there, although this won't be that important once the copy to clipboard button gets implemented (because the button does work in the browser).

tomasz1986 replied:
Marking resolved, because I've decided not to touch that script in this PR. The PR itself is still in the works, so please wait with reviewing until I mark it "ready for review" 🙂.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -37,8 +37,12 @@
+                <button type="button" class="btn btn-default" ng-click="copyToClipboard(currentDevice.deviceID)">
+                  <span class="fas fa-files-o"></span>&nbsp;<span class="ng-scope" translate>Copy</span>
+                </button>
+                <a class="btn btn-default" ng-click="copyToClipboard(currentDevice.deviceID)" href="mailto:?subject={{deviceName(currentDevice)}}%20has%20invited%20you%20to%20Syncthing&body={{deviceName(currentDevice)}}%20has%20invited%20you%20to%20Syncthing!%0D%0A%0D%0AYou%20can%20accept%20the%20invitation%20by%20adding%20a%20new%20device%20using%20the%20following%20ID.%20If%20unsure%20what%20to ...
```
`<button>` is impossible here, because `mailto` can be used inside `href` only. The copy to clipboard is a mistake (thanks for noticing!), but this is still a work in progress.

I'll see what I can then with moving the whole thing to JS.

tomasz1986 replied:
Moved everything to JS.

tomasz1986 commented on gui/default/syncthing/device/idqrModalView.html:
```diff
@@ -1,6 +1,14 @@
-    <div class="well well-sm text-monospace text-center" select-on-click>{{currentDevice.deviceID}}</div>
+    <div class="input-group">
+      <div class="well well-sm text-monospace form-control text-center">{{currentDevice.deviceID}}</div>
+      <div class="input-group-btn">
```
I've tried doing that, but it didn't really look nicely overall. I'm still planning to do more experiments with the layout.

tomasz1986 commented:
> Some GUI styling issues can probably be avoided entirely. If you describe what different behavior you'd like to achieve with the CSS, maybe I can help find a solution.

This is still a work in progress more or less. I'll try to fix the current issues first, and the come back with what's left and ask for more comments and input 🙂.

acolomb commented:
The email text sounds fine to me, except I don't know if "invited to Syncthing" conveys enough context.  Syncthing is not a place, and not the device is inviting, but the user sending the mail.  Suggestion:

`You are being invited to synchronize with '<device-name>' using Syncthing!`

Note the quotes around the device name.  Could also use "collaborate" or "connect" instead of "synchronize".

st-review commented:
:robot: *beep boop*

I'm going to close this pull request as it has been idle for more than 90 days.

This is not a rejection, merely a removal from the list of active pull requests that is periodically reviewed by humans. The pull request can be reopened when there is new activity, and merged once any remaining issues are resolved.

acolomb reviewed (changes requested):
Thought I'd have a look at the current draft status again.  Overall it looks nice, but I spotted some possible improvements.  One bigger request, please make this applicable to the local device as well, e.g. in the "Show QR" modal dialog.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3026,6 +3026,55 @@ angular.module('syncthing.core')
+                var textarea = document.createElement("textarea");
+                event.currentTarget.appendChild(textarea);
+                textarea.style.position = "fixed";
+                textarea.style.opacity = "0";
```
I will check those, but I did test a lot of different options. The whole copy to clipboard thing is a big mess when it comes to cross-browser compatibility, hence a few different methods are required. There's also a problem with the Bootstrap modal obstructing the view. I don't remember exactly now, but the fixed position may be required strictly in order to make the copy function work within the modal.

tomasz1986 replied:
After more testing, I can say that the element must be visible for the "copy to clipboard" to work, hence some hackery based on `opacity` is used as an alternative here. Of course, we're only talking about the third method, which relies on `document.queryCommandSupported`, as the other two don't require any temporary elements. However, out of the two other methods, the first has poor browser support (e.g. in won't work in the app on older versions of Android) while also requiring HTTPS (which isn't enforced in the Web GUI by default), and the second one is limited to IE only.

In other words, the third `document.queryCommandSupported` method is the most reliable, and it should actually work in all browsers, but at the same time it is also marked as "deprecated" (see https://developer.mozilla.org/docs/Web/API/Document/queryCommandSupported), so I think it's better to have all three for the sake of the widest compatibility.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3026,6 +3026,55 @@ angular.module('syncthing.core')
+        };
+
+        $scope.shareDeviceIdBy = function(channel) {
+            var body = $translate.instant("You have been invited to connect with {%name%} using Syncthing!", {name: $scope.currentDevice.name}) + "\r\n\r\n" + $translate.instant("Accept the invitation by adding a new device using the following ID.") + $translate.instant("If unsure what to do, continue reading below.") + "\r\n\r\n" + $scope.currentDevice.id + "\r\n\r\n" + $translate.instant("Synct ...
```
I was worried about making this a single, very long string, as then even a single change will require re-translating everything. I will have a look whether it could be reduced to a few larger, separate paragraphs though.

tomasz1986 replied:
I've modified the code, so that strings like the title, invitation, and footer are still separated, but the long description text is kept as a single entity, which should prevent situations, where we end up with English sentences right in between translated ones.

I don't really believe we should lump the whole email/sms text together though, as this will make it basically impossible to add any fixes later on (e.g. in a possibility that the "learn more" URL changes, etc.).

acolomb replied:
Looks better now, but please split once more on the first body line break / paragraph.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3026,6 +3026,55 @@ angular.module('syncthing.core')
+                var subject = $translate.instant("Invitation to connect with {%name%} using Syncthing", {name: $scope.currentDevice.name});
+                location.href = "mailto:?subject=" + encodeURIComponent(subject) + "&body=" + encodeURIComponent(body);
+            } else if (channel == 'sms') {
+                location.href = "sms://;?&body=" + encodeURIComponent(body);
```
> That's 4 SMS messages already.

I think it will actually be more when writing in non-Latin scripts. The whole body text should probably be reduced to a minimum. I will try to work on this later.

The `&` is required though. For the record, the sms code itself is based on https://stackoverflow.com/a/58131833.

tomasz1986 replied:
I have greatly reduced the SMS text to the required minimum. It takes exactly 146 characters now, but this is with the `{%name%}` placeholder. The actual device name is likely to be longer than that, but we should still have ~10 characters to spare. The main problem is the device ID itself, which normally uses 56 characters, so I've decided to strip the optional hyphens from it.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -14,6 +14,15 @@
+                </button>
+                <button type="button" class="btn btn-default" ng-click="shareDeviceIdBy('sms')" ng-disabled="editingDeviceNew() && !deviceEditor.deviceID.$valid">
+                  <span class="fas fa-comment-o"></span>&nbsp;<span translate>Share by SMS</span>
+                </button>
```
> Thought I'd have a look at the current draft status again. Overall it looks nice, but I spotted some possible improvements. One bigger request, please make this applicable to the local device as well, e.g. in the "Show QR" modal dialog.

Thank you for all the comments, although the current state of the code is still very far from complete. I planned to work on the local device dialog, but only after I've completed the other one, for the sake of simplicity.

tomasz1986 replied:
I think I've implemented the suggested changes. The text has been moved to tooltips, and the icons have been enlarged. The buttons themselves are now available both in the "Edit Device" modal, and also in the "Show QR" modal.

tomasz1986 commented on script/translate.go:
```diff
@@ -24,7 +24,7 @@ import (
 var attrRe = regexp.MustCompile(`\{\{\s*'([^']+)'\s+\|\s+translate\s*\}\}`)
 var attrReCond = regexp.MustCompile(`\{\{.+\s+\?\s+'([^']+)'\s+:\s+'([^']+)'\s+\|\s+translate\s*\}\}`)
-var jsRe = regexp.MustCompile(`\$translate.instant\("([^"]+)"\)`)
+var jsRe = regexp.MustCompile(`\$translate.instant\("([^"]+)"(,.*\)|\))`)
```
I will check this one, but to tell the truth, the current code here isn't really working. I was playing around with it, changing the patterns a few times, but still couldn't make it grab everything I wanted to, and finally I just had no time and left the code in the non-working state.

tomasz1986 replied:
The regex should be fixed now, but please have a look and let me know if you think it can be simplified more.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -14,6 +14,15 @@
               <div ng-if="!editingDeviceNew()" class="well well-sm form-control text-monospace" style="height: auto;" select-on-click>{{currentDevice.deviceID}}</div>
               <div class="input-group-btn">
+                <button type="button" class="btn btn-default" ng-click="copyToClipboard($event, currentDevice.deviceID)" ng-disabled="editingDeviceNew() && !deviceEditor.deviceID.$valid">
+                  <span class="fas fa-files-o"></span>&nbsp;<span translate>Copy to Clipboard</span>
```
Changed the icon to `fa-clone` now. Thanks for the tip! 🙂

acolomb commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -14,8 +14,17 @@
               <input ng-if="editingDeviceNew()" name="deviceID" id="deviceID" class="form-control text-monospace" type="text" ng-model="currentDevice.deviceID" required="" valid-deviceid list="discovery-list" aria-required="true" />
               <div ng-if="!editingDeviceNew()" class="well well-sm form-control text-monospace" style="height: auto;" select-on-click>{{currentDevice.deviceID}}</div>
               <div class="input-group-btn">
```
You could experiment with `input-group-addon` as it is used in `editShareTemplate.html` for the lock and eye indicators.  I think it looks more consistent especially with the dark theme.  Also please check if a separate `input-group-btn` or `input-group-addon` per button makes sense.  The `editShareTemplate.html` directive e.g. uses `<span>`, dunno if that makes a difference.

tomasz1986 replied:
It's a late reply, but I've just checked those. They're not great actually, as the `div`s and `span`s can't be focused with keyboard. Buttons, on the other hand, can, so they're much better in terms of accessibility in this regard.

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                        success = true;
+                    }
+                    catch (ex) {
+                        var error = $translate.instant("No email client found. Do you want to copy the email text to clipboard instead?");
```
This message always appears for me (on Firefox), right before the mail client shows up.  Same with the SMS error message, where the desktop browser is still asking for an application to handle sms:// links.

tomasz1986 replied:
Yep, after more testing and investigation, it seems that detecting `mailto` and `sms` protocol support is simply impossible, especially when external applications are involved. I did find one very hacky method that measured browser response time and based its decision on that (i.e. it concluded no support for the protocol if the browser responded immediately), but I don't think it's the right way to go, as the browser may respond slowly e.g. on low-end hardware, etc. too.

I've switched to a completely different approach now, which asks the user for confirmation using the built-in modals instead of JS alerts.

acolomb commented on script/translate.go:
```diff
@@ -24,7 +24,9 @@ import (
-var jsRe = regexp.MustCompile(`\$translate.instant\("([^"]+)"\)`)
+
+// Find both `$translate.instant("…")` and `$translate.instant("…",…)` in JS.
+var jsRe = regexp.MustCompile(`\$translate.instant\("((.+?))"(,|\))`)
```
Did you try my suggestion?  It would always look for the closing brace, even with the comma involved.

tomasz1986 replied:
I've changed the regex completely to catch all the different variants of `$translate.instant`. For testing, I cleared the English lang file completely and re-run the script from scratch, and it did find everything, i.e. the file came up exactly the same as before with no existing translations lost).

Please let me know if you think this should be simplified and/or done differently.

acolomb replied:
I do find the construct with alternative double `"` or single `'` quotes rather fragile, because it may start with one and end with the other style.  Starting from my suggestion, I tried just using two compiled regex variants and looping through them.  Much simpler and less confusing with mixed quotes, and the result is the same as in your test. Note that escaped quotes (`' this isn\'t nice'` or `"let's be \"nice\" today"`) don't work well with the translations at all, hence can be ignored.

Here's my code:

```go
var jsRe = []*regexp.Regexp{
	regexp.MustCompile(`\$translate\.instant\("([^"]+?)"(,.*\)|\))`),
	regexp.MustCompile(`\$translate\.instant\('([^']+?)'(,.*\)|\))`),
}
```

And below where it's used:

```go
			for s := bufio.NewScanner(fd); s.Scan(); {
				for _, re := range jsRe {
					for _, matches := range re.FindAllStringSubmatch(s.Text(), -1) {
						translation(matches[1])
					}
				}
			}
```

tomasz1986 replied:
Done.

acolomb commented:
I couldn't resist taking a look at the new wip state ;-)  Hope you can make use of at least the coding style comments.

tomasz1986 commented:
> I couldn't resist taking a look at the new wip state ;-) Hope you can make use of at least the coding style comments.

Thank you very much for all the comments, although you will probably waste some time this way, as I myself am still trying out different approaches 😉.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                              $scope.copyToClipboard(onclick, body);
+                        }
+                    }
+                    break;
```
Are you sure about this one? The other cases of `switch` found in `syncthingController.js` don't seem to use any blank lines like that.

acolomb replied:
The difference is that those other examples are only one or very few lines.  Just visually, the amount of code within this case justifies a blank line in between, be it in a `case` block or any other coherent larger part of a function.  My opinion obviously, but at least with a good reason ;-)

tomasz1986 replied:
I've tried adding blank lines, but they didn't look that great. The code used in each `case` has shrunken in size a bit in the meantime, so I don't think blank lines are really that much of a help now…

I personally don't find it difficult to distinguish between different `case`s here, but this obviously may be due to syntax highlighting and other visual cues that I use in the code editor, e.g.

![image](https://user-images.githubusercontent.com/5626656/166099729-0d91b336-5ab8-4c45-a928-9e550e5ccab1.png)

Do you think line breaks are still necessary at the moment?

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                        }
+                    }
+                    break;
+            }
```
Added `default:`s, but to be fair, most of the other cases of `switch` found in `syncthingController.js` don't seem to use it 😉.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                    body += '\r\n\r\n';
+                    body += $scope.currentDevice.deviceID.replace(/-/g, '');
+                    body += '\r\n\r\n';
+                    body += footer;
```
I think this needs to be `\n\n` still, as without the empty lines, everything feels very cramped.

acolomb replied:
Well, it's two more characters toward the 160 limit.  SMS isn't particularly suited for "pretty" messages, so let's optimize for brevity instead.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+        $scope.shareDeviceId = function(method) {
+            // Title and footer can be reused between email, SMS, and possibly
+            // other methods, hence we define them separately in the beginning.
+            var title = $translate.instant("{%name%} has shared their Syncthing device ID!", {name: $scope.currentDevice.name});
```
I've added a fallback to short device ID if the name is empty.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+                    // device ID. The current minimum length is around 150 chars,
+                    // but some room is required for longer sharing device names.
+                    var body = title;
+                    body += '\r\n\r\n';
```
Yep, as far as things go, the CRLFs seem to be only required in emails (by the spec), and not in SMS messages.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,95 @@ angular.module('syncthing.core')
+            if (navigator.clipboard && navigator.clipboard.writeText) {
+                // Default for modern browsers on localhost or HTTPS.
+                navigator.clipboard.writeText(text);
+                return alert(success);
```
I've changed the code following your second suggestion. However, do you think it's safe to default to `success` like that rather than default to `failure` and change the message to `success` only in the successful cases instead?

acolomb replied:
As long as this function is short enough to view all at once, yes it's safe.  It's easy enough to verify that each case either sets the failure message or is actually successful.  The suggested default was just considering which way needs less lines of code.

acolomb reviewed (changes requested):
Some more coding style comments and I think you have a rather technical perspective on the explanation texts.  Thus I tried to give some suggestions for more user-friendly and less technical strings.

I really like the new preview dialog overall, explaining what will (hopefully) happen.  It's one more click to achieve the end result, but guiding the user through the process might be worth it.  Just forget that the reason for it was "I can't detect whether it works", allowing shorter and less confusing instructions.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,120 @@ angular.module('syncthing.core')
                     address.indexOf('unixs://') == 0);
         };
+
+        $scope.shareDeviceConfirmationModal = function (type) {
```
Yep, I've changed it to `type` because of the very similar function used for the revert local additions confirmation. Do you think it's better to go with `method` even though the existing and very similar code uses `type`? I was just trying to be consistent here.

acolomb replied:
Yes.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,120 @@ angular.module('syncthing.core')
+                    // Ref: https://datatracker.ietf.org/doc/html/rfc5322
+                    params.subject = title;
+                    params.body = [
+                        $translate.instant("You have been invited to connect with {%devicename%} using Syncthing!", {devicename: deviceName}),
```
I don't really think "you are being invited` is any better here. Google gives way more results for "you have been invited" and "you've been invited" vs the counterparts, and the form itself is commonly used in invitations like that, e.g. https://i.pinimg.com/736x/df/d5/b5/dfd5b5f7ee59e155bf7f2afca7acdf7f.jpg or https://files.readme.io/4e1dd2d-Screen_Shot_2019-10-01_at_18.55.36.png.

acolomb replied:
Remember who the sender will be. Let's compare it with something like Dropbox for example. Their invitation mails are sent from their server, telling Randy Receiver that Sally Sender has started the action of inviting him.

In this case, it is Sally sending the message, so ideally it should say "Hello Randy, I am inviting you!" The action is just taking place and it is the initiator writing the message herself. Not a third party informing Randy about what has happened. Syncthing is different from typical cloud services in this regard, by design, and it's a Good Thing if the message reflects that difference.

tomasz1986 replied:
The whole message has changed a bit, and there's no "invited" anymore, so I'm marking this one resolved.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceDialogView.html:
```diff
@@ -0,0 +1,52 @@
+      <span translate>The following content will automatically be inserted into a message.</span>
+      <span ng-switch="shareDeviceParams.type">
+        <span ng-switch-when="email">
+          <span translate>You should then use your own email account to send it.</span>
+          <span translate>Proceed only if you have a registered email client.</span>
```
The `registered email client` is a Microsoft term. Windows literally tells you that if there's no registered handler for the email protocol. I assumed anyone anyone who'd ever used Windows in the last 30 years would be familiar with the term, but I'll try to make the sentence more accessible nevertheless 😉.

I'm not a fan of using the word `app` or `program` here though, because nowadays quite a lot of people have their Web browser (or rather Gmail specifically) registered as the default e-mail handler, which isn't exactly a separate application.

> What else should the user do?

If the "share" button doesn't work, the user is supposed to copy the content to clipboard, and then manually proceed to go to their e-mail and send it there. I think some kind of a warning is necessary here, because the OS/browser will often go wonky if there's no registered application and display some kind of a warning with a lot of technical jargon, which the user have no idea what to do with.

acolomb replied:
I agree that "program" or "application" is not well suited and too technical. But the term "app" has evolved into a widely known substitute for the "thing on a screen you interact with". Even Gmail in the browser can be called an app, and Syncthing's own GUI is another prime example.

If the browser gives a stupid error message, that's their fault and we try to set the right expectations beforehand. That's what I like about the new dialog. When the user sees unexpected or cryptic messages, it's very clearly like "OK this way doesn't work, let's go back one step and try differently". At that point anyone having mastered Syncthing to the point of recommending it to others should know basic things like how to forward a text displayed on screen to a friend via their usual messenger of choice.

My "what else should they do" comment was specifically targeted at the "You should..." sentence, which I found to just repeat what the user is already trying to do. The "proceed only..." warning sounds scary including the registration part, and leaving no recommendation for those who would rather not do anything like registering. So just tell them what should happen, let them see if it works (because we cannot reliably predict) and describe the alternative. Hence my suggested wording.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceDialogView.html:
```diff
@@ -0,0 +1,52 @@
+          <span translate>Proceed only if you have a registered SMS client.</span>
+        </span>
+      </span>
+      <span translate>Alternatively, copy the content to clipboard and send it manually yourself.</span>
```
Do you think it would be fine to remove the word "clipboard" from the button text entirely and just leave "copy" instead? It can still stay in the code, of course, but the truth is that a typical computer user has zero idea what "clipboard" actually means.

acolomb replied:
I guess so, neither Discourse nor Github seems to use it based on my quick checks ;-)

tomasz1986 replied:
Renamed to just "copy". It's still called "copyToClipboard" in the code itself.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,120 @@ angular.module('syncthing.core')
+            }
+        }
+
+        $scope.showTooltip = function (newTooltip) {
```
Alerts are usually a no-go in general, mainly because they require an immediate action preventing the user from interacting with the document completely, but also they are extremely abused by spammy sites in order to prevent the user from leaving the site, etc., so many people have grown to hate them in principle. That's why after more testing I've decided that it'd be much better to use the modal notifications instead.

However, this particular case is different, as the purpose of the function is simply to either show a temporary tooltip or replace the existing one, so that the user knows that their action has been successful (or not), .e.g. `Copy to Clipboard` becomes `Copied to Clipboard!`, etc. I will add the proper explanation, so that everything should be clear about its purpose.

acolomb replied:
That's alright. I just meant you should put a comment on the function what its purpose and mechanism is. Maybe some less generic naming would help also. Tool tips are used in many places, so this should make obvious how it's different from a regular tool tip.

tomasz1986 replied:
Explanation added. Please let me know if you've got a better, and maybe more specific, name in mind. The function itself can be used to add a temporary tooltip to any element though, hence I've used a more comprehensive name for it.

acolomb replied:
`showTemporaryTooltip`?

It's not quite clear in your comment what "afterwards" refers to.  I think you mean after the temporary one is shown, the data for an event-driven tooltip is restored.

tomasz1986 replied:
Renamed to `afterwards` and replaced "afterwards" with a better explanation.

uok commented:
You should also take a look at https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share which enables native sharing, e.g. on Android it opens actions like "share with contact on Signal" 

I have added it to my own projects with multiple buttons e.g. "email", "facebook" and "share". I then show/hide whatever buttons are supported/not supported.
```
if (navigator.share) { 
  // sharing supported - hide all buttons except "share"
}
else {
  // sharing not supported - hide share button, keep fallback buttons
}
```
It works really well!

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,120 @@ angular.module('syncthing.core')
+                        footer
+                    ].join('\n\n');
+                    break;
+                default:
```
I've removed the `default`s for now now to make the code follow the existing convention… `default` itself is marked "optional" on https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/switch too, so I guess it's really probably not needed here (especially if it's just empty).

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3077,6 +3077,120 @@ angular.module('syncthing.core')
+                var deviceName = $scope.currentDevice.name;
+            } else {
+                var deviceName = $scope.deviceShortID(deviceID);
+            }
```
To be specific, I meant: `$scope.deviceName($scope.currentDevice)`, which should give you the desired result.

tomasz1986 replied:
Done.

bt90 commented:
@tomasz1986 are you still working on this? Would love to see this merged.

tomasz1986 commented:
> @tomasz1986 are you still working on this? Would love to see this merged.

Yeah, still planning to get back to this one sometime in the near future. Thanks for asking 🙂.

As for the the mobile-specific code suggested above, I'll probably leave it to a separate PR that would possibly follow this one, as the PR is already quite extensive, and I'm also worried that if I start to deep dive into the mobile sharing, I'll never be able to finish it.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceDialogView.html:
```diff
@@ -0,0 +1,52 @@
+      </span>
+      <span translate>Alternatively, copy the content to clipboard and send it manually yourself.</span>
+    </p>
+    <hr>
```
I've tried to remove the `hr`, but to be honest, everything looks to cramped for me without it. I think it looks better when there's clear distinction between the help message above, and the actual message content below.

tomasz1986 replied:
Done. The email variant wouldn't need the HR either, but it's just the margin between `p` and `hr5` that's just too narrow by default 😕. The styling comes from Bootstrap though.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceDialogView.html:
```diff
@@ -0,0 +1,52 @@
+      <span translate>Alternatively, copy the content to clipboard and send it manually yourself.</span>
+    </p>
+    <hr>
+    <div ng-if="shareDeviceParams.type == 'email'">
```
Done.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceDialogView.html:
```diff
@@ -0,0 +1,52 @@
+        ><button type="button" class="btn btn-default pull-right" ng-click="copyToClipboard(shareDeviceParams.subject)" tooltip data-original-title="{{ 'Copy to Clipboard' | translate }}">
+          <span class="fa fa-clone"></span>
+        </button
+      ></pre>
```
Done, although it's not ideal, but you can't do

```
        </button>
      </pre>
```
as the new line and spaces between the two will be present (and copied) in the content.

acolomb replied:
Yes, that.

tomasz1986 replied:
Done.

tomasz1986 commented:
This is now ready for review. I'm really sorry for taking so long 🙁.

There're quite a few code changed and additions here and there, so I wouldn't be surprised if a lot of that still needs fixes and improvements.

I've also added a few screenshots to the first post.

tomasz1986 commented on gui/default/syncthing/device/idqrModalView.html:
```diff
@@ -1,6 +1,19 @@
+          <span class="fa fa-lg fa-envelope-o"></span>
+        </button>
+        <button type="button" class="btn btn-default" ng-click="shareDeviceConfirmationModal('sms')" ng-disabled="editingDeviceNew() && !deviceEditor.deviceID.$valid" tooltip data-original-title="{{ 'Share by SMS' | translate }}">
+          <span class="fa fa-lg fa-comments-o"></span>
```
Done.

acolomb commented:
Great to see you working on this again :-) I'll need some time for another review, because of the many details.

A general comment about PR review: Could you please not mark conversations resolved after replying on them? GitHub's interface is unfortunately pretty convoluted and hard to keep track of follow-ups. But if everything's resolved, it gets very hard to verify what was changed, or find the context of the last reply (which I usually read in the notification email).

On a related note, I'd welcome slightly more descriptive commit messages than always `wip`, and rebasing only when absolutely necessary.  Instead of looking at the overall result over and over, my workflow is very change-centric, trying to keep track of which comments were addressed and how. Having single, quickly identifiable commits on top of the last reviewed state helps with that. Thank you.

tomasz1986 commented:
Thanks for still keeping an eye on the PR 😉. There're quite a lot changes since the last time. I've tried to simplify the code as much as possible, but I'm sure more needs to be done still.

> Could you please not mark conversations resolved after replying on them?

I'm very sorry for that, but the issue is there were so many of them that if I hadn't marked them "resolved", I would've lost track on what I'd already fixed and what I hadn't 🙁. Also, many became somewhat "obsolete", with the related code completely removed. I'll try to just leave them open from now.

> On a related note, I'd welcome slightly more descriptive commit messages than always `wip`, and rebasing only when absolutely necessary.

To be honest, this may be difficult, as I've already got used so much to the `wip` commits… That's the style I was told PRs were normally done here in the very beginning, so I've just followed the example since then. I think you're probably the only one that does descriptive commits here. I'll definitely try not to rebase though! The only problem with that is that there's no way other than rebase if I want to modify the main commit message, or is there?

acolomb commented:
> ... if I hadn't marked them "resolved", I would've lost track on what I'd already fixed and what I hadn't

Yeah, that's what I meant, GitHub's PR interface is not really good to keep track of things, neither for the reviewer, nor the author.  It should be more like the issues list, but internal to the PR.  The "conversations" idea is fine in general, just really hard to follow visually.

> To be honest, this may be difficult, as I've already got used so much to the `wip` commits… That's the style I was told PRs were normally done here in the very beginning, so I've just followed the example since then.

IMO, it's fine for your own, incremental work on the code, but less so for addressing review comments.  It basically makes the commit shortlog completely useless.  It's true that any elaborate commit messages in between will get concatenated or lost when squash-merging, but during review they are still useful to follow what happens.  Just a few words are usually enough.

> The only problem with that is that there's no way other than rebase if I want to modify the main commit message, or is there?

True, but in the end the final commit message is edited when squash-merging.  The individual commit messages and, since [recently](https://forum.syncthing.net/t/pr-commit-message-from-title/18437), the PR title will be part of the suggestion.  The accepted practice here seems to be that only little info ends up in the commit, but the PR reference carries all the context.  So it's usually not worth writing and re-writing the "perfect" commit message by rebasing.  It's really up to the person doing the merge.

tomasz1986 commented:
> I will look at the new state soon and get back to you on the actual changes.

Just to possibly make things easier, the JS code is mostly the same as before, just with a few minor fixes and naming changes here and there (and improved comments). The HTML has changed quite a bit, plus the whole `idqrModalView.html` page has basically been newly added.

> The accepted practice here seems to be that only little info ends up in the commit, but the PR reference carries all the context.

I'm not really sure about that, as all my commits recently merged do have their full commit message in them. To be exact here, I believe that's the case when @calmh is the one, who merges them 😉.

Is it also confusing if a rebase is done only to edit the commit message while leaving everything else intact?

calmh commented:
It depends. When merging GitHub will suggest a concatenation of all the commit messages in the branch. If that looks like shit I generally remove the crap. If that leaves the initial commit message entirely outdated because the target moved during development, I typically remove that too.

If the suggested commit message looks decent with just minor touchups I usually retain it.

acolomb commented:
I've opened #8542 to isolate the changes needed for the translation script.

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3124,6 +3124,132 @@ angular.module('syncthing.core')
+                        title + ":",
+                        deviceID.replace(/-/g, ''),
+                        footer
+                    ].join('\n\n');
```
The message has only three-lines, we can spare the two extra newlines to gain some more headroom.

```suggestion
                    ].join('\n');
```

acolomb replied:
Thanks.  In addition to saving the two characters, it looks much better now, at least in my SMS app view.

tomasz1986 replied:
Eh, I still think it's kind of cramped, especially on smaller screens, where the ID wraps along multiple lines, but I've checked SMS that I receive from banks and such, and they also put a lot of content without any empty lines there, so I guess that's the way things are normally done 🤔.

acolomb commented on gui/default/syncthing/device/shareDeviceIDDialogView.html:
```diff
@@ -0,0 +1,39 @@
+      </span>
+      <span translate>You can also copy and paste the content into the message manually.</span>
+    </p>
+    <hr>
```
Move down one line, so it only applies for email.

tomasz1986 replied:
Done.

acolomb commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -13,9 +13,18 @@
-                <button type="button" class="btn btn-default" data-toggle="modal" data-target="#idqr" ng-disabled="editingDeviceNew() && !deviceEditor.deviceID.$valid">
-                  <span class="fas fa-qrcode"></span>&nbsp;<span translate>Show QR</span>
+              <div id="shareDeviceIdButtons" class="input-group-btn">
+                <button type="button" class="btn btn-default" ng-click="copyToClipboard(currentDevice.deviceID)" ng-disabled="editingDeviceNew() && !deviceEditor.deviceID.$valid" tooltip data-original-title="{{ 'Copy' | translate }}">
```
Somehow the "Copied!" tooltip re-appears for me when closing and re-opening this modal.  Maybe it should be cleared when closing the dialog?

tomasz1986 replied:
I can't reproduce this one. Can you describe the steps in detail? I've tested in Chromium and Firefox, but to no avail. I did find out that the script seems broken in Palemoon (which likely means that it doesn't work in older versions of Firefox either), but that's a different problem.

acolomb replied:
No idea, I can't reproduce it anymore now.

acolomb commented on gui/default/syncthing/device/shareDeviceIDDialogView.html:
```diff
@@ -0,0 +1,39 @@
+      <pre style="word-break: normal; white-space: pre-wrap;">{{shareDeviceIdParams.subject}}<button
+        type="button" class="btn btn-default pull-right" ng-click="copyToClipboard(shareDeviceIdParams.subject)" tooltip data-original-title="{{ 'Copy' | translate }}">
+          <span class="fa fa-clone"></span>
+        </button
+      ></pre>
```
```suggestion
      </button></pre>
```

tomasz1986 replied:
Done.

acolomb commented on gui/default/syncthing/device/shareDeviceIDDialogView.html:
```diff
@@ -0,0 +1,39 @@
+    <pre style="word-break: normal; white-space: pre-wrap;">{{shareDeviceIdParams.body}}<button
+      type="button" class="btn btn-default pull-right" ng-click="copyToClipboard(shareDeviceIdParams.body)" tooltip data-original-title="{{ 'Copy' | translate }}">
+        <span class="fa fa-clone"></span>
+      </button
+    ></pre>
```
```suggestion
    </button></pre>
```

tomasz1986 replied:
Done.

acolomb commented on script/translate.go:
```diff
@@ -24,7 +24,10 @@ import (
+
+// Find both $translate.instant("…") and $translate.instant("…",…) in JS.
+// Consider single quote variants too.
+var jsRe = regexp.MustCompile(`\$translate.instant\((?:"|')(.+?)(?:"|')(?:, *?\{.+\})*?\)`)
```
See #8542.

acolomb reviewed (changes requested):
This is starting to look really good.  I believe you fixed most of my older comments, so let's focus on the remaining ones here.  I did read through the whole diff again and also did some local UI testing on Firefox (PC, Linux) and on Android (Firefox and Chrome).

What doesn't work for me is the SMS template.  On the PC, it correctly opens my GSConnect interface to send the SMS via my mobile, but fails with a dialog where I should "select a device".  Dunno, that might be a fragile construct overall, so might be disregarded.  On Android though, it does correctly open the SMS app, but says something about "Reply to this short code not possible" (translated from German).  The intended recipient looks like it was parsed as `//`, so maybe just prefix with `sms:` instead of `sms://`?

![Screenshot_20220915-003344](https://user-images.githubusercontent.com/6996887/190274876-af6bc97c-ccf7-42c3-a522-443dea89a41b.png)

acolomb commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3124,6 +3124,132 @@ angular.module('syncthing.core')
+            // Title and footer can be reused between different sharing
+            // methods, hence we define them separately before the body.
+            var title = $translate.instant('Syncthing device ID for "{%devicename%}"', {devicename: deviceName});
+            var footer = $translate.instant("Learn more at {%url%}.", {url: "https://syncthing.net"});
```
Drop the period, it might interfere with the URL.  And it's one more character saved for SMS :-)

tomasz1986 commented:
> What doesn't work for me is the SMS template. On the PC, it correctly opens my GSConnect interface to send the SMS via my mobile, but fails with a dialog where I should "select a device". Dunno, that might be a fragile construct overall, so might be disregarded. On Android though, it does correctly open the SMS app, but says something about "Reply to this short code not possible" (translated from German). The intended recipient looks like it was parsed as `//`, so maybe just prefix with `sms:` (works for me) instead of `sms://`?

This is interesting. I will need to do some proper testing here. One problem is that I've only got access to Samsung Android and maybe LineageOS, so my hands are tied up a little.

I don't think we can drop `//` though, as `sms://` acts as the protocol here (like `http://`). From what I've read, it may actually be impossible to support all different cases with this one, as some messaging apps simply don't like the fact that there's no pre-defined recipient.

Out of curiosity, what kind of Android (and SMS app) is this?

bt90 commented:
https://www.rfc-editor.org/rfc/rfc5724#section-2.5

`sms:` is the correct prefix

tomasz1986 commented:
Oops. Thank you very much for the link, and I must admit that I got fooled then :cold_sweat:.

Will make the changes soon.

acolomb commented:
My testing was on a Google Pixel 4a with up-to-date stock Android 13.

Would be good to get some confirmation from an iOS user.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceIDDialogView.html:
```diff
@@ -0,0 +1,39 @@
+<modal id="share-device-id-dialog" status="warning" icon="{{shareDeviceIdParams.icon}}" heading="{{shareDeviceIdParams.heading}}" large="{{ shareDeviceIdParams.method == 'email' ? 'yes' : 'no' }}" closeable="yes">
```
I did actually rename it, but apparently Git for Windows isn't capable of detecting case only filename changes. I've now used the trick from https://stackoverflow.com/questions/17683458/how-do-i-commit-case-sensitive-only-filename-changes-in-git/40307511#40307511, which did the job.

acolomb replied:
Glad you found a way. I guess another possibility would have been GitHub's online editor, though I have never used it myself.

tomasz1986 commented on gui/default/syncthing/device/shareDeviceIDDialogView.html:
```diff
@@ -0,0 +1,39 @@
+          <span translate>Your SMS app should open to let you choose the recipient and send it from your own number.</span>
+        </span>
+      </span>
+      <span translate>You can also copy and paste the content into the message manually.</span>
```
I actually used "content" on purpose, so that we don't repeat the word "text" over and over. Do you think it's better to use "text" here as well?

acolomb replied:
I found it a bit inconsistent while looking at the dialog, and thought it might be more confusing to talk about "content" here for the first time.  Better stick to the wording used right above ("The following text ...")

And while we're at it, I guess "paste the text into _a_ message manually" would fit better. When done manually, it's not really the same message mentioned before which supposedly opens automatically.

tomasz1986 replied:
Done.

tomasz1986 commented on gui/default/assets/lang/lang-en.json:
```diff
@@ -450,6 +463,7 @@
    "When adding a new folder, keep in mind that the Folder ID is used to tie folders together between devices. They are case sensitive and must match exactly between all devices.": "When adding a new folder, keep in mind that the Folder ID is used to tie folders together between devices. They are case sensitive and must match exactly between all devices.",
    "Yes": "Yes",
    "Yesterday": "Yesterday",
+   "You can also copy and paste the content into the message manually.": "You can also copy and paste the content into the message manually.",
```
I actually like to keep translations bundled together with their commit, so if it needs to be reverted later for some reason, the unused translations can easily be removed with it as well. This wouldn't be necessary if the Go translate script simply refreshed the whole `json` file on each run (which I'd personally prefer, because it'd allow the person to see whether any possible changes to the script haven't broken anything), but that's a completely different story…

acolomb replied:
Good point about the revert.  I don't feel strongly about this, just wanted to give a hint about not forgetting such changes while the PR is still a moving target.  Fine with me, feel free to resolve this.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -3124,6 +3124,132 @@ angular.module('syncthing.core')
+                    params.subject = title;
+                    params.body = [
+                        $translate.instant('"{%devicename%}" has shared their Syncthing device ID!', {devicename: deviceName}),
+                        $translate.instant("To connect with them, add a new device using the ID below."),
```
Done.

tomasz1986 commented:
I think I've addressed all the comments. I've also updated the screenshots to reflect the current state. I'm still going to test the SMS thing on real Android devices. I'm also going to give the QR code attachment a try, but I fail to progress with it in a reasonable amount of time, I'll probably leave this part for a different commit later on. The same goes for the native share functionality (which actually applies to all types of different systems, even including desktop ones like Windows).

acolomb commented on gui/default/syncthing/device/shareDeviceIdDialogView.html:
```diff
@@ -12,21 +12,21 @@
+    <pre style="word-break: normal; white-space: pre-wrap;">{{shareDeviceIdParams.body}}<button type="button" class="btn btn-default pull-right" ng-click="copyToClipboard(shareDeviceIdParams.body)" tooltip data-original-title="{{ 'Copy' | translate }}">
+      <span class="fa fa-clone"></span>
+    </button></pre>
+    <p>In addition, you can download and attach the ID encoded as QR code as well!</p>
+    <div>
+      <a href="qr/?text={{currentDevice.deviceID}}" target="_blank" download="{{currentDevice.deviceID}}.png"><img class="img-thumbnail" ng-src="qr/?text={{currentDevice.deviceID}}" height="109" width="109" alt="{{'QR code' | translate}}"></a>
+    </div>
```
Hm... I don't really see the purpose of this.  When sending via SMS, it's going to cause a costly (at least with German providers) MMS instead.  For email, I doubt many people would succeed in scanning a QR code from an email attachment (I wouldn't know how easily).

If there is a legitimate use case to send the QR code as image, it can already be downloaded from the `idqrModalView`.  Whoever needs that will also know how to attach the downloaded file to a message somewhere.  You could make it easier by linking to the downloadable file from the large QR code image, which would save a right click (long press) and a click on "Save image" or similar.  Not sure it's worth it though.

tomasz1986 replied:
This is highly country-dependent, but there are places, where MMS costs the same or almost the same as normal SMS messages. However, you're right that scanning an image located on the device itself would be problematic. Ideally, the QR scanner that's built into the Synching app should have this functionality, but right now it's only capable of using the camera to take a picture of the QR.

The "downloadable" button next to the large QR image isn't a bad idea, I think. This is because non-techy people actually don't know that they can right click and save images.

acolomb replied:
What I meant was to make the image itself clickable (a link), not another separate button.  Just like how your thumbnail works, just on the original.

tomasz1986 replied:
Done.

acolomb reviewed (changes requested):
Great, I think we're really getting this thing polished up and I really like how it makes sharing device IDs much quicker.  Thanks for heeding my suggestions about commit titles and review.  It really helps me a lot to keep the review manageable.

IMHO the last addition with the downloadable QR is too much, let's keep that for a separate discussion / PR until deemed useful.

Oh, and I re-tested the SMS thing on my Android, works flawlessly now.

tomasz1986 commented:
> Oh, and I re-tested the SMS thing on my Android, works flawlessly now.

We really need to have someone test this on the iPhone…

acolomb commented:
> That's great to hear. Now, we really need to have someone test this on the iPhone…

Maybe @xor-gate can help?  Steps for testing:
1. Build from latest PR commit.
2. Run on your build machine (Mac / PC).
3. Configure GUI listen address to 0.0.0.0.
4. Open the PC / Mac's LAN IP address in the iOS browser, e.g. `http://192.168.0.123:8384`.
5. Click on the short ID for "This Device".
6. Click the SMS sharing button.
7. Confirm with the "Share" button.
8. Pre-filled SMS should open and ask for a recipient.

xor-gate commented:
> > That's great to hear. Now, we really need to have someone test this on the iPhone…
> 
> Maybe @xor-gate can help? Steps for testing:
> 
> 1. Build from latest PR commit.
> 2. Run on your build machine (Mac / PC).
> 3. Configure GUI listen address to 0.0.0.0.
> 4. Open the PC / Mac's LAN IP address in the iOS browser, e.g. `http://192.168.0.123:8384`.
> 5. Click on the short ID for "This Device".
> 6. Click the SMS sharing button.
> 7. Confirm with the "Share" button.
> 8. Pre-filled SMS should open and ask for a recipient.

Dear @acolomb,

I indeed own a iPhone. To be exact iPhone 7 Plus with iOS 15.6.1. I do have Chrome for iOS installed. And Safari of course.
I have it planned to test this PR for next week. As i'm in the middle of someones house moving project.

Thanks for the clear instruction steps.

I will give feedback when I'm done.

xor-gate commented:
This is how it looks at iOS with Safari. The receipt is now filled in with weird ';'. 

![IMG_3877](https://user-images.githubusercontent.com/1050166/190918751-b7ff5511-9156-4d82-a575-6bb3e545317a.PNG)
![IMG_3878](https://user-images.githubusercontent.com/1050166/190918753-8b2c35b7-136c-4303-a3ac-4cec0f0ed102.PNG)
![IMG_3879](https://user-images.githubusercontent.com/1050166/190918756-8b47be05-7d00-45a5-88fe-101b80993e39.PNG)

bt90 commented:
I think iOS uses a different body separator:

```
sms:phonenumber&body=sometext
```

xor-gate commented:
It seems now the To field is empty (instead of `;` and the cursor/selection of the input box is now correct:

<img width="375" alt="PNG image" src="https://user-images.githubusercontent.com/1050166/191057164-fa978fa5-163e-442d-bbb3-2c2966ee3d7f.png">

xor-gate commented:
On macOS 12.3.1 it also works fine for SMS/message. Also the email share works as expected:

<img width="605" alt="Screenshot 2022-09-19 at 17 43 39" src="https://user-images.githubusercontent.com/1050166/191058147-61bd4290-8db7-42d7-9e6a-2d9c6c15f394.png">

tomasz1986 commented:
That was quick 🙂. It seems we've got macOS and iOS covered then.

The current iteration also works fine for me here on Samsung Android.

acolomb reviewed (dismissed):
Confirmed to work on my Google Pixel (Android 13) as well.

Just out of curiosity, is there similar functionality in recent Windows versions? Can e.g. Microsoft's "Surface" tablets with cellular modem send SMS?  (Shouldn't block this PR though when we have the two biggest mobile OSes covered.)

tomasz1986 commented:
> Just out of curiosity, is there similar functionality in recent Windows versions? Can e.g. Microsoft's "Surface" tablets with cellular modem send SMS? (Shouldn't block this PR though when we have the two biggest mobile OSes covered.)

Judging by what people write online, the answer is "yes", at least on some devices, but I haven't used any Windows tablet with a modem built in, so I don't really know any details. In theory, the link should work there too.

xor-gate commented:
@tomasz1986 because nobody did merge after approval the feature branch is now unable to be merged. Could you sync the changes back so the branch is mergeable?

tomasz1986 commented:
> @tomasz1986 because nobody did merge after approval the feature branch has a merge conflict and is now unable to be merged. Could you sync the changes back so the branch is mergeable?

Done! Thanks for giving a hint, as I completely forgot about checking the older PRs for conflicts :flushed:.

xor-gate commented:
@acolomb i'm not involved in the syncthing daemon development. Who is responsible for merging so it can be shipped with the v1.22.2 release window?

acolomb commented:
I was hesitant to merge because no other maintainer had commented much. And I don't see myself making last calls on features without an opinion from one of them.

xor-gate commented:
> I was hesitant to merge because no other maintainer had commented much. And I don't see myself making last calls on features without an opinion from one of them.

Totaly agree with this. But LGTM.

acolomb commented:
After no objections, let's go ahead and try pouring this over the brave RC users...
